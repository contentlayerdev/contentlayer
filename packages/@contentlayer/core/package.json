{
  "name": "@contentlayer/core",
  "version": "0.2.6-dev.8",
  "type": "module",
  "exports": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "./dist/**/*.{js,ts,map}",
    "./src",
    "./package.json"
  ],
  "scripts": {
    "test": "echo No tests yet"
  },
  "peerDependencies": {
    "date-fns": "2.x",
    "esbuild": "^0.12.1 || 0.13.x || 0.14.x",
    "markdown-wasm": "1.x"
  },
  "peerDependenciesMeta": {
    "date-fns": {
      "optional": true
    },
    "esbuild": {
      "optional": true
    },
    "markdown-wasm": {
      "optional": true
    }
  },
  "dependencies": {
    "@contentlayer/utils": "workspace:*",
    "camel-case": "^4.1.2",
    "comment-json": "^4.2.2",
    "date-fns": "2.x",
    "esbuild": "^0.12.1 || 0.13.x || 0.14.x",
    "gray-matter": "^4.0.3",
    "mdx-bundler": "^9.0.0",
    "rehype-stringify": "^9.0.3",
    "remark-frontmatter": "^4.0.1",
    "remark-parse": "^10.0.1",
    "remark-rehype": "^10.1.0",
    "source-map-support": "^0.5.21",
    "type-fest": "^2.12.2",
    "unified": "^10.1.2"
  },
  "devDependencies": {
    "@types/source-map-support": "^0.5.4",
    "markdown-wasm": "^1.2.0"
  }
}
