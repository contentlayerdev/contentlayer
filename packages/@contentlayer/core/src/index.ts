export * from './generation/generate-dotpkg.js'
export * from './generation/generate-types.js'
export * from './DataCache.js'
export * from './data-types.js'
export * from './cwd.js'
export * from './extensions.js'
export * from './gen.js'
export * from './errors.js'
export * from './getConfig/index.js'
export * from './runMain.js'
export * from './markdown/markdown.js'
export * from './markdown/mdx.js'
export * from './markdown/unified.js'
export * from './plugin.js'
export * from './schema/index.js'
export * from './ArtifactsDir.js'
export * from './validate-tsconfig.js'
