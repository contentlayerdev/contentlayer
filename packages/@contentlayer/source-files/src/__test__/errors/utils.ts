import type * as core from '@contentlayer/core'
import type { RelativePosixFilePath } from '@contentlayer/utils'
import { capitalizeFirstLetter, singleItem, unknownToRelativePosixFilePath } from '@contentlayer/utils'
import faker from 'faker'
import * as path from 'node:path'

import { FetchDataError } from '../../errors/index.js'

export const makeSchemaDef = (): core.SchemaDef => {
  const TypeA: core.DocumentTypeDef = {
    _tag: 'DocumentTypeDef',
    name: 'TypeA',
    description: undefined,
    fieldDefs: [],
    computedFields: [],
    extensions: {},
    isSingleton: false,
  }
  const TypeB: core.DocumentTypeDef = {
    _tag: 'DocumentTypeDef',
    name: 'TypeB',
    description: undefined,
    fieldDefs: [],
    computedFields: [],
    extensions: {},
    isSingleton: false,
  }
  const schemaDef: core.SchemaDef = { hash: '', nestedTypeDefMap: {}, documentTypeDefMap: { TypeA, TypeB } }
  return schemaDef
}

const generateFakeFilePath = (extension = 'md'): RelativePosixFilePath =>
  singleItem(path.join('docs', faker.system.commonFileName(extension))).map(unknownToRelativePosixFilePath).item

export const makeErrors = (
  countRecord: Partial<Record<FetchDataError.FetchDataError['_tag'], number>>,
): FetchDataError.FetchDataError[] => {
  const errors: FetchDataError.FetchDataError[] = []

  faker.seed(123)

  doNTimes(countRecord.CouldNotDetermineDocumentTypeError, () => {
    const documentFilePath = generateFakeFilePath()
    errors.push(new FetchDataError.CouldNotDetermineDocumentTypeError({ typeFieldName: 'type', documentFilePath }))
  })

  doNTimes(countRecord.NoSuchDocumentTypeError, () => {
    const documentFilePath = generateFakeFilePath()
    const documentTypeName = capitalizeFirstLetter(faker.hacker.noun())
    errors.push(new FetchDataError.NoSuchDocumentTypeError({ documentFilePath, documentTypeName }))
  })

  doNTimes(countRecord.UnexpectedError, () => {
    const documentFilePath = generateFakeFilePath()
    errors.push(
      new FetchDataError.UnexpectedError({
        documentFilePath,
        error: new Error(`Some problem happened: ${faker.hacker.phrase()}`),
      }),
    )
  })

  doNTimes(countRecord.ComputedValueError, () => {
    const documentFilePath = generateFakeFilePath()
    const error = new Error(`Some problem happened: ${faker.hacker.phrase()}`)
    errors.push(new FetchDataError.ComputedValueError({ documentFilePath, error }))
  })

  doNTimes(countRecord.ExtraFieldDataError, (index) => {
    const documentFilePath = generateFakeFilePath()
    const documentTypeName = capitalizeFirstLetter(faker.hacker.noun())
    errors.push(
      new FetchDataError.ExtraFieldDataError({
        documentFilePath,
        documentTypeName,
        extraFieldEntries: [index === 0 ? ['someKey', 'someVal'] : ['someOtherKey', 42]],
      }),
    )
  })

  doNTimes(countRecord.MissingRequiredFieldsError, () => {
    const documentFilePath = generateFakeFilePath()
    const documentTypeName = capitalizeFirstLetter(faker.hacker.noun())
    const fieldDef: core.FieldDef = {
      type: 'string',
      name: 'someField',
      default: undefined,
      description: undefined,
      isRequired: true,
      isSystemField: false,
    }
    errors.push(
      new FetchDataError.MissingRequiredFieldsError({
        documentFilePath,
        documentTypeName,
        fieldDefsWithMissingData: [fieldDef],
      }),
    )
  })

  return errors
}

const doNTimes = (n_: number | undefined, fn: (i: number) => void): void => {
  const n = n_ ?? 0
  for (let i = 0; i < n; i++) {
    fn(i)
  }
}
