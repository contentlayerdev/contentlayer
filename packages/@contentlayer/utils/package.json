{
  "name": "@contentlayer/utils",
  "version": "0.2.7-dev.9",
  "type": "module",
  "exports": {
    "./package.json": {
      "import": "./package.json"
    },
    ".": {
      "import": "./dist/index.js"
    },
    "./effect": {
      "import": "./dist/effect/index.js"
    },
    "./effect/Tracing/Enable": {
      "import": "./dist/effect/Tracing/Enable.js"
    },
    "./node": {
      "import": "./dist/node/index.js"
    }
  },
  "types": "./dist/index.d.ts",
  "typesVersions": {
    "*": {
      "package.json": [
        "./package.json"
      ],
      "effect": [
        "./dist/effect"
      ],
      "effect/Tracing/Enable": [
        "./dist/effect/Tracing/Enable"
      ],
      "node": [
        "./dist/node"
      ]
    }
  },
  "files": [
    "./dist/**/*.{js,ts,map}",
    "./src",
    "./package.json"
  ],
  "scripts": {
    "test": "echo No tests yet"
  },
  "peerDependenciesMeta": {
    "@effect-ts/core": {
      "optional": true
    },
    "@effect-ts/otel": {
      "optional": true
    },
    "@effect-ts/otel-node": {
      "optional": true
    }
  },
  "dependencies": {
    "@effect-ts/core": "^0.58.0",
    "@effect-ts/otel": "^0.13.0",
    "@effect-ts/otel-exporter-trace-otlp-grpc": "^0.13.0",
    "@effect-ts/otel-sdk-trace-node": "^0.13.0",
    "@opentelemetry/api": "~1.0.3",
    "@opentelemetry/core": "~1.0.0",
    "@opentelemetry/exporter-trace-otlp-grpc": "~0.27.0",
    "@opentelemetry/resources": "~1.0.0",
    "@opentelemetry/sdk-trace-base": "~1.0.0",
    "@opentelemetry/sdk-trace-node": "~1.0.0",
    "@opentelemetry/semantic-conventions": "~1.0.0",
    "chokidar": "^3.5.3",
    "hash-wasm": "^4.9.0",
    "inflection": "^1.13.2",
    "oo-ascii-tree": "^1.63.2",
    "ts-pattern": "^4.0.5",
    "type-fest": "^2.18.0"
  },
  "devDependencies": {
    "@types/inflection": "^1.13.0"
  }
}
