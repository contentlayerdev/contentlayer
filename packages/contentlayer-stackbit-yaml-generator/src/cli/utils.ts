import YAML from 'yaml'

export const toYamlString = (json: any): string => YAML.stringify(json, { aliasDuplicateObjects: false })
