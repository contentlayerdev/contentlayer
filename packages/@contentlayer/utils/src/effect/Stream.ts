import { pipe } from '@effect-ts/core'
import * as Chunk from '@effect-ts/core/Collections/Immutable/Chunk'
import * as Tuple from '@effect-ts/core/Collections/Immutable/Tuple'
import * as T from '@effect-ts/core/Effect'
import * as S from '@effect-ts/core/Effect/Experimental/Stream'
import * as E from '@effect-ts/core/Either'

export * from '@effect-ts/core/Effect/Experimental/Stream'

export const streamTapSkipFirst =
  <R1, E1, O, X>(f: (o: O) => T.Effect<R1, E1, X>) =>
  <R, E>(stream: S.Stream<R, E, O>): S.Stream<R & R1, E | E1, O> =>
    pipe(
      stream,
      S.mapAccumEffect(0, (x, o) =>
        T.gen(function* (_) {
          if (x > 0) {
            yield* _(f(o))
          }
          return Tuple.tuple(x + 1, o)
        }),
      ),
    )

/** Note this function doesn't currently work if the first value is a `E.left` value */
export const tapSkipFirstRight =
  <R1, R2, E1, EE1, E2, A1>(f: (o: A1) => T.Effect<R2, E2, any>) =>
  (stream: S.Stream<R1, E1, E.Either<EE1, A1>>): S.Stream<R1 & R2, E1 | E2, E.Either<EE1, A1>> =>
    pipe(
      stream,
      S.zipWithIndex,
      S.tap(({ tuple: [val, index] }) => (index === 0 || E.isLeft(val) ? T.succeed(null) : f(val.right))),
      S.map(({ tuple: [val] }) => val),
    )

export const tapRight =
  <R1, R2, E1, EE1, A1>(f: (o: A1) => T.Effect<R2, never, unknown>) =>
  (stream: S.Stream<R1, E1, E.Either<EE1, A1>>): S.Stream<R1 & R2, E1, E.Either<EE1, A1>> =>
    pipe(
      stream,
      S.tap((val) => (E.isLeft(val) ? T.succeed(null) : f(val.right))),
    )

export const tapLeft =
  <R1, R2, E1, EE1, A1>(f: (e: EE1) => T.Effect<R2, never, unknown>) =>
  (stream: S.Stream<R1, E1, E.Either<EE1, A1>>): S.Stream<R1 & R2, E1, E.Either<EE1, A1>> =>
    pipe(
      stream,
      S.tap((val) => (E.isLeft(val) ? f(val.left) : T.succeed(null))),
    )

export const tapRightEither =
  <R1, R2, E1, EE1, EE2, A1>(f: (o: A1) => T.Effect<R2, never, E.Either<EE2, unknown>>) =>
  (stream: S.Stream<R1, E1, E.Either<EE1, A1>>): S.Stream<R1 & R2, E1, E.Either<EE1 | EE2, A1>> =>
    pipe(
      stream,
      S.tap((val) => (E.isLeft(val) ? T.succeed(null) : f(val.right))),
    )

export const startWith =
  <A2>(...values: A2[]) =>
  <R1, E1, A1>(stream: S.Stream<R1, E1, A1>): S.Stream<R1, E1, A2 | A1> =>
    S.merge_(stream, S.fromChunk(Chunk.from(values)))

export const startWithRight =
  <A2>(value: A2) =>
  <R1, E1, A1>(stream: S.Stream<R1, never, E.Either<E1, A1>>): S.Stream<R1, never, E.Either<E1, A2 | A1>> =>
    // TODO: Is this correct? Technically, I think this inserts the "init" _after_ any events -> race condition?
    // how about `S.merge_(S.fromIterable([E.right(value)]), stream)`
    S.merge_(stream, S.fromIterable([E.right(value)]))

export const chainMapEitherRight =
  <R2, E2, EE2, A2, A1>(mapRight: (a1: A1) => S.Stream<R2, E2, E.Either<EE2, A2>>) =>
  <R1, E1, EE1>(stream: S.Stream<R1, E1, E.Either<EE1, A1>>): S.Stream<R1 & R2, E1 | E2, E.Either<EE1 | EE2, A2>> => {
    return S.chain_(
      stream,
      E.fold(
        (_left) => stream as any,
        (right) => mapRight(right),
      ),
    )
  }

export const chainSwitchMapEitherRight =
  <R2, E2, EE2, A2, A1>(mapRight: (a1: A1) => S.Stream<R2, E2, E.Either<EE2, A2>>) =>
  <R1, E1, EE1>(stream: S.Stream<R1, E1, E.Either<EE1, A1>>): S.Stream<R1 & R2, E1 | E2, E.Either<EE1 | EE2, A2>> => {
    return S.chainParSwitch<R2, E2, E.Either<EE1, A1>, E.Either<EE2, A2>>(
      E.fold(
        (_left) => stream as any,
        (right) => mapRight(right),
      ),
      1,
    )(stream)
  }

export const mapEffectEitherRight =
  <R2, E2, A2, A1>(mapRight: (a1: A1) => T.Effect<R2, never, E.Either<E2, A2>>) =>
  <R1, E1>(stream: S.Stream<R1, never, E.Either<E1, A1>>): S.Stream<R1 & R2, never, E.Either<E1 | E2, A2>> => {
    return S.mapEffect_(
      stream,
      E.fold(
        (left) => T.succeed(E.leftW<E1 | E2, A2>(left)),
        (right) => mapRight(right),
      ),
    )
  }

export const mapEitherLeft =
  <EE2, EE1>(mapLeft: (e1: EE1) => EE2) =>
  <R1, E1, A1>(stream: S.Stream<R1, E1, E.Either<EE1, A1>>): S.Stream<R1, E1, E.Either<EE2, A1>> => {
    return S.map_(stream, E.mapLeft(mapLeft))
  }

export const mapEitherRight =
  <A2, A1>(mapRight: (a1: A1) => A2) =>
  <R1, E1, EE1>(stream: S.Stream<R1, E1, E.Either<EE1, A1>>): S.Stream<R1, E1, E.Either<EE1, A2>> => {
    return S.map_(stream, E.map(mapRight))
  }
