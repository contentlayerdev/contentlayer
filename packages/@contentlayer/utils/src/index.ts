export * from './string.js'
export * from './guards.js'
export * from './object/index.js'
export * from './tracing.js'
export * from './promise.js'
export * from './hash.js'
export * from './single-item.js'
export * from './file-paths.js'
export * as base64 from './base64.js'
export * from './tracing-effect/index.js'

export { AsciiTree } from 'oo-ascii-tree'
export * as pattern from 'ts-pattern'
import { Tagged } from '@effect-ts/core/Case'
// inflection is a CJS module, so we need to import it as default export
import inflection from 'inflection'

export { inflection }

export const recRemoveUndefinedValues = (val: any): void => {
  if (Array.isArray(val)) {
    val.forEach(recRemoveUndefinedValues)
  } else if (typeof val === 'object') {
    Object.keys(val).forEach((key) => {
      if (val[key] === undefined) {
        delete val[key]
      } else {
        recRemoveUndefinedValues(val[key])
      }
    })
  }
}

export const partition = <T, TLeft extends T>(
  arr: T[],
  isLeft: (_: T) => _ is TLeft,
): [TLeft[], Exclude<T, TLeft>[]] => {
  return arr.reduce(
    (acc, el) => {
      if (isLeft(el)) {
        acc[0]!.push(el)
      } else {
        acc[1]!.push(el as Exclude<T, TLeft>)
      }
      return acc
    },
    [[], []] as [TLeft[], Exclude<T, TLeft>[]],
  )
}

export const not =
  <T, TGuarded extends T>(guard: (_: T) => _ is TGuarded) =>
  (el: T): el is Exclude<T, TGuarded> =>
    !guard(el)

export const errorToString = (error: any) => {
  const stack = process.env.CL_DEBUG ? error.stack : undefined
  const str = error.toString()
  const stackStr = stack ? `\n${stack}` : ''
  if (str !== '[object Object]') return str + stackStr

  return JSON.stringify({ ...error, stack }, null, 2)
}

export const capitalizeFirstLetter = (str: string): string => str.charAt(0).toUpperCase() + str.slice(1)

/**
 * Use this to make assertion at end of if-else chain that all members of a
 * union have been accounted for.
 */
export function casesHandled(x: never): never {
  throw new Error(`A case was not handled for value: ${JSON.stringify(x)}`)
}

export function notImplemented(msg?: string): never {
  throw new Error(`Not yet implemented ${msg}`)
}

export type Thunk<T> = () => T

export const unwrapThunk = <T>(_: T | (() => T)): T => {
  if (typeof _ === 'function') {
    return (_ as any)()
  } else {
    return _
  }
}

export class RawError extends Tagged('RawError')<{ readonly error: unknown }> {}

export const isReadonlyArray = <T>(_: any): _ is ReadonlyArray<T> => Array.isArray(_)

export function assertNever(_: any): never {
  throw new Error(`assertNever: This should never happen ${JSON.stringify(_)}`)
}

export const asMutableArray = <T>(arr: ReadonlyArray<T>): T[] => arr.slice()
