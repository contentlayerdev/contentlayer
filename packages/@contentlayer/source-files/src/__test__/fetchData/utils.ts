import type { HasCwd } from '@contentlayer/core'
import * as core from '@contentlayer/core'
import { provideCwd } from '@contentlayer/core'
import {
  provideDummyTracing,
  unknownToAbsolutePosixFilePath,
  unknownToRelativePosixFilePath,
} from '@contentlayer/utils'
import type { HasClock, HasConsole, OT } from '@contentlayer/utils/effect'
import { pipe, provideTestConsole, T, These } from '@contentlayer/utils/effect'

import type { HasDocumentTypeMapState } from '../../fetchData/DocumentTypeMap.js'
import { provideDocumentTypeMapState } from '../../fetchData/DocumentTypeMap.js'
import { testOnly_makeContentTypeMap, testOnly_makefilePathPatternMap } from '../../fetchData/index.js'
import { makeCacheItemFromFilePath } from '../../fetchData/makeCacheItemFromFilePath.js'
import type { DocumentTypes } from '../../index.js'
import { makeSource } from '../../index.js'

export const runTest = async ({
  documentTypes,
  contentDirPath: contentDirPath_,
  relativeFilePath: relativeFilePath_,
}: {
  documentTypes: DocumentTypes
  contentDirPath: string
  relativeFilePath: string
}) => {
  const eff = T.gen(function* ($) {
    const relativeFilePath = unknownToRelativePosixFilePath(relativeFilePath_)
    const contentDirPath = unknownToAbsolutePosixFilePath(contentDirPath_)
    const esbuildHash = 'not-important-for-this-test'

    const source = yield* $(T.tryPromise(() => makeSource({ contentDirPath, documentTypes })))
    const coreSchemaDef = yield* $(source.provideSchema(esbuildHash))

    const documentTypeDefs = (Array.isArray(documentTypes) ? documentTypes : Object.values(documentTypes)).map((_) =>
      _.def(),
    )
    const filePathPatternMap = testOnly_makefilePathPatternMap(documentTypeDefs)
    const contentTypeMap = testOnly_makeContentTypeMap(documentTypeDefs)

    const options: core.PluginOptions = {
      date: undefined,
      markdown: undefined,
      mdx: undefined,
      fieldOptions: core.defaultFieldOptions,
      disableImportAliasWarning: false,
    }

    const cache = yield* $(
      pipe(
        makeCacheItemFromFilePath({
          relativeFilePath,
          contentDirPath,
          coreSchemaDef,
          filePathPatternMap,
          options,
          previousCache: undefined,
          contentTypeMap,
        }),
        These.effectToEither,
      ),
    )

    return cache
  })

  return runMain(eff)
}

const runMain = async <E, A>(
  eff: T.Effect<OT.HasTracer & HasClock & HasCwd & HasConsole & HasDocumentTypeMapState, E, A>,
) => {
  const logMessages: string[] = []
  const result = await pipe(
    eff,
    provideTestConsole(logMessages),
    provideDocumentTypeMapState,
    provideCwd,
    provideDummyTracing,
    T.runPromise,
  )

  return { logMessages, result }
}
