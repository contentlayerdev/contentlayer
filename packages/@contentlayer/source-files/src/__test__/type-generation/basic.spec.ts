import { renderTypes } from '@contentlayer/core'
import { provideJaegerTracing } from '@contentlayer/utils'
import { pipe, provideConsole, T } from '@contentlayer/utils/effect'
import { expect, test } from 'vitest'

import type { DocumentTypes } from '../../index.js'
import { makeSource } from '../../index.js'
import { defineDocumentType } from '../../schema/defs/index.js'

const renderTypeSource = async (documentTypes: DocumentTypes) => {
  const esbuildHash = 'not-important-for-this-test'
  const schemaDef = await pipe(
    T.tryPromise(() => makeSource({ documentTypes, contentDirPath: '' })),
    T.chain((source) => source.provideSchema({ esbuildHash })),
    provideJaegerTracing('contentlayer-cli'),
    provideConsole,
    T.runPromise,
  )

  const typeSource = renderTypes({
    schemaDef,
    generationOptions: {
      sourcePluginType: 'local',
      options: {
        fieldOptions: { bodyFieldName: 'body', typeFieldName: 'type' },
        markdown: undefined,
        mdx: undefined,
        date: undefined,
        disableImportAliasWarning: false,
      },
    },
  })

  return typeSource
}

// TODO rewrite test for gendotpkg
test('generate-types: simple schema', async () => {
  const TestPost = defineDocumentType(() => ({
    name: 'TestPost',
    filePathPattern: `**/*.md`,
    fields: {
      title: {
        type: 'string',
        description: 'The title of the post',
        required: true,
      },
      date: {
        type: 'date',
        description: 'The date of the post',
        required: true,
      },
    },
    computedFields: {
      slug: { type: 'string', resolve: (_) => _._id.replace('.md', '') },
    },
  }))

  const typeSource = await renderTypeSource([TestPost])

  expect(typeSource).toMatchSnapshot()
})

test('generate-types: simple schema with optional fields', async () => {
  const TestPost = defineDocumentType(() => ({
    name: 'TestPost',
    filePathPattern: `**/*.md`,
    fields: {
      title: {
        type: 'string',
        description: 'The title of the post',
        required: true,
      },
      date: {
        type: 'date',
        description: 'The date of the post',
      },
    },
    computedFields: {
      slug: { type: 'string', resolve: (_) => _._id.replace('.md', '') },
    },
  }))

  const typeSource = await renderTypeSource([TestPost])

  expect(typeSource).toMatchSnapshot()
})

test('generate-types: references with embedded schema', async () => {
  const Post = defineDocumentType(() => ({
    name: 'Post',
    fields: {
      title: { type: 'string', required: true },
      author: {
        type: 'reference',
        of: Person,
        required: true,
        embedDocument: true,
      },
    },
  }))

  const Person = defineDocumentType(() => ({
    name: 'Person',
    fields: {
      name: { type: 'string', required: true },
    },
  }))

  const typeSource = await renderTypeSource([Post, Person])

  expect(typeSource).toMatchSnapshot()
})
