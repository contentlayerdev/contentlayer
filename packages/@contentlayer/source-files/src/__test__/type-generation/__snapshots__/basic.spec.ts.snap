// Vitest Snapshot v1

exports[`generate-types: references with embedded schema 1`] = `
"// NOTE This file is auto-generated by Contentlayer

import type { Markdown, MDX, ImageFieldData } from 'contentlayer/core'
import * as Local from 'contentlayer/source-files'

export { isType } from 'contentlayer/client'

// export type Image = string
export type { Markdown, MDX }

/** Document types */
export type Person = {
  /** File path relative to \`contentDirPath\` */
  _id: string
  _raw: Local.RawDocumentData
  type: 'Person'
  name: string
  /** Markdown file body */
  body: Markdown

}

export type Post = {
  /** File path relative to \`contentDirPath\` */
  _id: string
  _raw: Local.RawDocumentData
  type: 'Post'
  title: string
  author: Person
  /** Markdown file body */
  body: Markdown

}  

/** Nested types */
  

/** Helper types */

export type AllTypes = DocumentTypes | NestedTypes
export type AllTypeNames = DocumentTypeNames | NestedTypeNames

export type DocumentTypes = Person | Post
export type DocumentTypeNames = 'Person' | 'Post'

export type NestedTypes = never
export type NestedTypeNames = never


export interface ContentlayerGenTypes {
  documentTypes: DocumentTypes
  documentTypeMap: DocumentTypeMap
  documentTypeNames: DocumentTypeNames
  nestedTypes: NestedTypes
  nestedTypeMap: NestedTypeMap
  nestedTypeNames: NestedTypeNames
  allTypeNames: AllTypeNames
}

declare global {
  interface ContentlayerGen extends ContentlayerGenTypes {}
}

export type DocumentTypeMap = {
  Person: Person
  Post: Post
}

export type NestedTypeMap = {

}

 "
`;

exports[`generate-types: simple schema 1`] = `
"// NOTE This file is auto-generated by Contentlayer

import type { Markdown, MDX, ImageFieldData } from 'contentlayer/core'
import * as Local from 'contentlayer/source-files'

export { isType } from 'contentlayer/client'

// export type Image = string
export type { Markdown, MDX }

/** Document types */
export type TestPost = {
  /** File path relative to \`contentDirPath\` */
  _id: string
  _raw: Local.RawDocumentData
  type: 'TestPost'
  /** The title of the post */
  title: string
  /** The date of the post */
  date: string
  /** Markdown file body */
  body: Markdown
  slug: string
}  

/** Nested types */
  

/** Helper types */

export type AllTypes = DocumentTypes | NestedTypes
export type AllTypeNames = DocumentTypeNames | NestedTypeNames

export type DocumentTypes = TestPost
export type DocumentTypeNames = 'TestPost'

export type NestedTypes = never
export type NestedTypeNames = never


export interface ContentlayerGenTypes {
  documentTypes: DocumentTypes
  documentTypeMap: DocumentTypeMap
  documentTypeNames: DocumentTypeNames
  nestedTypes: NestedTypes
  nestedTypeMap: NestedTypeMap
  nestedTypeNames: NestedTypeNames
  allTypeNames: AllTypeNames
}

declare global {
  interface ContentlayerGen extends ContentlayerGenTypes {}
}

export type DocumentTypeMap = {
  TestPost: TestPost
}

export type NestedTypeMap = {

}

 "
`;

exports[`generate-types: simple schema with optional fields 1`] = `
"// NOTE This file is auto-generated by Contentlayer

import type { Markdown, MDX, ImageFieldData } from 'contentlayer/core'
import * as Local from 'contentlayer/source-files'

export { isType } from 'contentlayer/client'

// export type Image = string
export type { Markdown, MDX }

/** Document types */
export type TestPost = {
  /** File path relative to \`contentDirPath\` */
  _id: string
  _raw: Local.RawDocumentData
  type: 'TestPost'
  /** The title of the post */
  title: string
  /** The date of the post */
  date?: string | undefined
  /** Markdown file body */
  body: Markdown
  slug: string
}  

/** Nested types */
  

/** Helper types */

export type AllTypes = DocumentTypes | NestedTypes
export type AllTypeNames = DocumentTypeNames | NestedTypeNames

export type DocumentTypes = TestPost
export type DocumentTypeNames = 'TestPost'

export type NestedTypes = never
export type NestedTypeNames = never


export interface ContentlayerGenTypes {
  documentTypes: DocumentTypes
  documentTypeMap: DocumentTypeMap
  documentTypeNames: DocumentTypeNames
  nestedTypes: NestedTypes
  nestedTypeMap: NestedTypeMap
  nestedTypeNames: NestedTypeNames
  allTypeNames: AllTypeNames
}

declare global {
  interface ContentlayerGen extends ContentlayerGenTypes {}
}

export type DocumentTypeMap = {
  TestPost: TestPost
}

export type NestedTypeMap = {

}

 "
`;
